### ISSUE TITLE
WordNetLemmatizer not properly lemmatizing some words

### ISSUE THREAD

**Issue by gorj-tessella** [nltk/nltk#2567]
Some words are lemmatized improperly, due to picking the smallest possible lemma: 

```py
lemmatizer = WordNetLemmatizer()
lemmatizer.lemmatize('dose', 'n') # returns "dose"
lemmatizer.lemmatize('doses', 'n') # returns "dos"
wordnet._morphy('doses', 'n') # returns ["dose", "dos"]
wordnet.morphy('doses', 'n') # returns "dose"
```

**Comment by tomaarsen** [comment]
@gorj-tessella 
I've written the following program to quickly get an overview of how `WordNetLemmatizer` and `morphy` compare.
```python
from nltk.stem.wordnet import WordNetLemmatizer
from nltk.corpus import wordnet as wn
import csv

# Amount of times the two lemmatizers resulted in the same lemma
identical = 0
# Total amount of accepted test cases
total = 0
# The times wordnet or morphy had the better result respectively
wordnet_wins = set()
morphy_wins = set()

wnl = WordNetLemmatizer()
with open("noun.csv", "r", errors='replace') as f:
    reader = csv.reader(f)
    for line in reader:
        singular = line[0]
        plurals = line[1:] # There might be multiple plurals
        
        for plural in plurals:
            # Lemmatize according to WordNetLemmatizer and morphy
            wn_l = wnl.lemmatize(plural, pos="n")
            m_l = wn.morphy(plural, pos="n")
            
            # Ignore if morphy is unable to lemmatize
            if m_l is not None:
                if wn_l != m_l:
                    # If wordnet is right and morphy is not:
                    if wn_l == singular:
                        wordnet_wins.add(plural)
                    # If morphy is right and wordnet is not:
                    if m_l == singular:
                        morphy_wins.add(plural)
                
                else:
                    identical += 1
                total += 1

# In case there are alternate spellings, add them to "tie"
# and remove them from the individual ones
tie = wordnet_wins.intersection(morphy_wins)
wordnet_wins -= tie
morphy_wins -= tie

breakpoint()
```
(Python 3.7+, Python 3.5+ onwards if you remove the `breakpoint()`)

This program will go through a file `"noun.csv"`, which I downloaded from https://github.com/djstrong/nouns-with-plurals/blob/master/noun.csv. 
Each of the ~142000 plurals were lemmatized both by `WordNetLemmatizer` and `morphy`. All cases where `morphy` was unable to find a lemma were discarded. For the remaining cases it's checked whether there's a difference in results. If there is, the method with the correct lemma (if any) had that plural added to its respective set (i.e. `wordnet_wins` or `morphy_wins`). If there was no difference, `identical` is incremented. Lastly `total` is incremented for each non-discarded lemma.
Afterward, all plurals both in `wordnet_wins` and in `morphy_wins` (e.g. if a plural has multiple possible lemmas, and each of the methods produce one of those valid lemmas) will be added to `tie`, and those plurals will be removed from `wordnet_wins` and `morphy_wins`.

---

## Results
Out of more than 142000 plurals, morphy was only able to lemmatize 31613, as it requires the lemma to be a known word according to Wordnet, and many of the plurals in the list are multiple words (`snow ploughs`) or simply not real words (`σ-finite measures`). Out of these 31613 test cases, the two methods resulted in the same lemma 30987 times, or ~97.59% of the time. 

The remaining times there was a difference. The interesting part here is finding out which one was accurate more often.

---

For each of these plurals, `WordNetLemmatizer` is right and `morphy` is wrong.
```python
['abs', 'acoustics', 'acres', 'aesthetics', 'affairs', 'aides', 'aids', 'allies', 'aloes', 'alps', 'ambages', 'amenities', 'anagrams', 'anas', 'ancients', 'anklets', 'ans', 'antipodes', 'antitrades', 'aras', 'archives', 'ares', 'arms', 'as', 'ascomycetes', 'assets', 'assizes', 'baas', 'balusters', 'banks', 'baptists', 'barrels', 'bars', 'basics', 'basidiomycetes', 'bbs', 'beads', 'beatniks', 'beats', 'bellows', 'bends', 'billings', 'bitters', 'bleachers', 'blinks', 'bloomers', 'blues', 'boards', 'bounds', 'bowels', 'bowls', 'boxcars', 'boxers', 'braces', 'brakes', 'breakers', 'brethren', 'bridges', 'briefs', 'brits', 'brooks', 'bunches', 'buns', 'burnouses', 'burns', 'buttocks', 'callas', 'canaries', 'candlepins', 'canticles', 'cascades', 'ceres', 'chains', 'chambers', 'channels', 'charades', 'checkers', 'cheviots', 'chilblains', 'chips', 'chives', 'circumstances', 'clams', 'clappers', 'classics', 'cleaners', 'cleats', 'cleavers', 'clews', 'clocks', 'clutches', 'cobblers', 'cocos', 'coevals', 'collards', 'comforts', 'comics', 'commons', 'communications', 'compliments', 'conditions', 'congratulations', 'conserves', 'contemporaries', 'contents', 'contras', 'conveniences', 'cords', 'corduroys', 'corrections', 'cos', 'costs', 'cows', 'crabs', 'craps', 'credentials', 'creeps', 'creepy-crawlies', 'crossroads', 'cs', 'customs', 'cyclopes', 'darts', 'das', 'davys', 'days', 'debs', 'deeds', 'descendants', 'deserts', 'devices', 'dialectics', 'diggings', 'digs', 'dippers', 'dominoes', 'dominos', 'dos', 'doubles', 
'dozens', 'draughts', 'drawers', 'dregs', 'duckpins', 'duds', 'dumas', 'dumplings', 'dumps', 'dunkers', 'dynamics', 'eas', 'eggs', 'elements', 'elves', 'ethics', 'eyeglasses', 'eyes', 'falls', 'fas', 'fatigues', 'feelings', 'fesses', 'fields', 'fifties', 'finances', 'findings', 'fives', 'fixings', 'flaps', 'flats', 'flies', 'folks', 'follies', 'followers', 'forties', 'fries', 'fumes', 'fundamentals', 'funds', 'funnies', 'gasteromycetes', 'gates', 'gens', 'giblets', 'glassworks', 'goldfields', 'gospels', 'graphics', 'greaves', 'greens', 'gripes', 'grits', 'groats', 'grotesqueries', 'groves', 'guts', 'gyps', 'hackles', 'hands', 'hanks', 'harmonics', 'hays', 'heaps', 'heavens', 'heaves', 'highlands', 'hindquarters', 'hippies', 'hipsters', 'hives', 'hooks', 
'hoops', 'hops', 'horseshoes', 'hours', 'huaraches', 'humans', 'hurdles', 'hymeneals', 'hysterics', 'indris', 'ins', 'ironsides', 'isometrics', 'jacks', 'jackstraws', 'jaspers', 'jimmies', 'jitters', 'johns', 'judges', 'junkers', 'khakis', 'knobkerries', 'knuckles', 'ks', 'lamentations', 'lancers', 'lashings', 'lats', 'laurels', 'laws', 'leaders', 'lees', 'leftovers', 'legs', 'leotards', 'letters', 'liabilities', 'limbers', 'loads', 'lodgings', 'logos', 'loins', 'loos', 'losses', 'lots', 'lowlands', 'lyons', 'majors', 'manes', 'manners', 'marbles', 'marches', 'marines', 'marks', 'mars', 'marshals', 'masses', 'masters', 'mates', 'maths', 'maulers', 'mays', 'means', 'mechanics', 'medlars', 'megrims', 'methodists', 'metrics', 'mills', 'minors', 'minutes', 'mnemonics', 'mods', 'monas', 'mopes', 'morals', 'mores', 'myxomycetes', 'najas', 'names', 'nerves', 'ninepins', 'nones', 'nothings', 'numbers', 'occasions', 'oddments', 'operations', 'optics', 'organs', 'outskirts', 'oxen', 'pants', 'parks', 'parsons', 'parts', 'peanuts', 'pickings', 'piles', 'pinches', 'plaudits', 'pliers', 'plyers', 'polemics', 'polls', 'pooves', 'porcupines', 
'prelims', 'premises', 'primates', 'privates', 'proceedings', 'profits', 'propaedeutics', 'prophets', 'props', 'proverbs', 'provisions', 'psalms', 'pyrites', 'quadratics', 'queens', 'quoits', 'raffles', 'rafts', 'rails', 'rastas', 'rates', 'receipts', 'relations', 'reserves', 'rings', 'roads', 'rockers', 'rooms', 'roots', 'rounders', 'ruddles', 'rudiments', 'sales', 'sauternes', 
'saxes', 'scads', 'scopes', 'scores', 'scots', 'scraps', 'scruples', 'seats', 'sellers', 'sens', 'services', 'sessions', 'settlings', 'shakers', 'shambles', 'shears', 'shekels', 'shirtsleeves', 'shoes', 'shorts', 'shucks', 'silks', 'sills', 'silversides', 'singles', 'sis', 'skinheads', 'skittles', 'skivvies', 'slews', 'slops', 'snips', 'snuffers', 'sops', 'sos', 'soviets', 'spareribs', 'specs', 'spectacles', 'spillikins', 'spirits', 'splinters', 'spots', 'sprinkles', 'sprites', 'stacks', 'staggers', 'stairs', 'stakes', 'stalls', 'stamina', 'stations', 'stays', 'steps', 'stigmata', 'stockholdings', 'stops', 'straits', 'stripes', 'sweetbreads', 'tabernacles', 'tactics', 'tails', 'talks', 'taps', 'taxis', 'tears', 'teens', 'tenpins', 'terms', 'testudines', 
'things', 'thirties', 'threads', 'throes', 'tigers', 'tons', 'tours', 'transactions', 'trappings', 'trembles', 'trimmings', 'troglodytes', 'troops', 'tropics', 'trumpets', 'trunks', 'tums', 'twenties', 'twins', 'values', 'vapors', 'velours', 'vespers', 'viands', 'vibes', 'victuals', 'viewers', 'waders', 'wads', 'wages', 'wales', 'watts', 'ways', 'weeds', 'wells', 'whiskers', 'windows', 'wings', 'winnings', 'wits', 'woods', 'words', 'workings', 'yaws', 'years', 'yips']
```
In total, `WordNetLemmatizer` performs better than `morphy` on 470 plurals.

---

For each of these plurals, `morphy` is right and `WordNetLemmatizer` is wrong.
```python
['beanies', 'colors', 'corps', 'corpses', 'coss', 'cruses', 'dies', 'doses', 'gendarmeries', 'grazes', 'grounds', 'kurus', 'morses', 'motives', 'muses', 'paraleipses', 'ploughmen', 'raves', 'reeves', 'reverses', 'roomies', 'ruses', 'senses', 'serves', 'sharpies', 'species', 'stogies', 'taeniae', 'touracos', 'uses', 'vases', 'weirdies']
```
In total, `morphy` performs better than `WordNetLemmatizer` on 32 plurals, including your `doses`.

---

And for each of these plurals both methods return a different but valid alternative spelling.
```python
['adzes', 'aeries', 'alexanders', 'annexes', 'aunties', 'battle-axes', 'bennies', 'blintzes', 'bogies', 'bolshies', 'booties', 'broadaxes', 'caddies', 'cartouches', 'cookies', 'coolies', 'cowries', 'crosses', 'darkies', 'dearies', 'dickies', 'doggies', 'dogies', 'eyries', 'faeries', 'floozies', 'goonies', 'grannies', 'hankies', 'hoagies', 'honkies', 'innings', 'junkies', 'kelpies', 'lenses', 'links', 'loonies', 'marquises', 'meanies', 'mews', 'mollies', 'oraches', 'organdies', 'panties', 'pas', 'pavises', 'pickaxes', 'pinkies', 'pixies', 'poleaxes', 'punkies', 'quickies', 'reveries', 'scrubs', 'shingles', 'sises', 'smoothies', 'softies', 'statistics', 'stymies', 'thrips', 'townies']
```
In total, there are 62 occurrences of both methods returning a different but valid alternative spelling.

---

In conclusion, though choosing the shortest lemma (like `WordNetLemmatizer`) does sometimes produce problems (e.g. for `"doses"`), in the grand majority of times it is *better* than simply taking the first option from `_morphy` like `morphy` does. This can be seen for e.g. `"abs"`, where `_morphy` returns `["abs", "ab"]`.

So, this is not something that should be changed. `WordNetLemmatizer` *should* pick the smallest lemma.

- Tom Aarsen

**Comment by caiw** [comment]
Also where `WordNetLemmatizer` is wrong:
`"possesses"` -> `"posse"`
`"ramesses"` -> `"ram"`
`"james "`-> `"jam"`
`"iss"` -> `"i"`

**Comment by Acervans** [comment]
Also `"riding"` and `"rides"` -> `"rid"`

**Comment by novalis** [comment]
`anchoresses` -> `anchor`
`siped` -> `sip`
`trapes` -> `trap`
`askeses` -> `ask`
`bibless` -> `bible`
`bowses` -> `bow`
`carses` -> `car`
`cates` -> `cat`
`cateresses` -> `cater` (`pos=v` -- it should return nothing, since `cateresses` is not a verb)
`chowses` -> `chow`
`hydrases` -> `hydra`
`idlesses` -> `idle`
`marses` -> `mars` (or `mar`, with `pos=v`)
`pareses` -> `par`
`replicases` -> `replica`
`semises` -> `semi`
`tared` -> `tar`
`taring` -> `tar`
`tootses` -> `toot`
`torqueses` -> `torque`

**Comment by ekaf** [comment]
Like most linguistic tools, lemmatizers are not perfect, their accuracy is an open research problem, and more research is needed in order to achieve perfect lemmatization. This suggests to draw a distinction between open research problems vs. software issues. 

(Edited) So, as long as we think that NLTK correctly implements the best known algorithms, there might not be a real software issue. But thanks to [@nezda's comment below ](https://github.com/nltk/nltk/issues/2567#issuecomment-1872328404), we now know that NLTK's implementation cannot be correct.

However, it is also well known that _Princeton morphy_ is overly permissive (it "over-generates"). This is usually considered a convenient feature for analysis, and not a bug. For example, the Princeton _wn_ program recognizes _cates_ as _cat_, and _siped_ as _sip_. It is a known feature of the original _morphy_ algorithm, and fixing it would mean implementing a different algorithm.

This suggests fixing only those errors in NLTK's _morphy_ implementation, which do not stem from the original Princeton algorithm.

**Comment by novalis** [comment]
A lookup table would be 100% accurate (for cases that are included), and would not require new research. It would require some sweat, but we have the start of such a table right here in this GitHub issue.

**Comment by nezda** [comment]
Seems like the original WordNet's [Morphy](https://wordnet.princeton.edu/documentation/morphy7wn) implementation doesn't have many of these bugs and maybe these are in fact bugs in this implementation? For example:
* http://wordnetweb.princeton.edu/perl/webwn?s=possesses&sub=Search+WordNet&o2=&o0=1&o8=1&o1=1&o7=&o5=&o9=&o6=&o3=&o4=&h= - no "possesses" →  "posse"
* http://wordnetweb.princeton.edu/perl/webwn?s=ramesses&sub=Search+WordNet&o2=&o0=1&o8=1&o1=1&o7=&o5=&o9=&o6=&o3=&o4=&h=0000000000000000 - no "ramesses" → "ram"
* http://wordnetweb.princeton.edu/perl/webwn?s=james&sub=Search+WordNet&o2=&o0=1&o8=1&o1=1&o7=&o5=&o9=&o6=&o3=&o4=&h=0 - no "james " → "jam"
* http://wordnetweb.princeton.edu/perl/webwn?s=iss&sub=Search+WordNet&o2=&o0=1&o8=1&o1=1&o7=&o5=&o9=&o6=&o3=&o4=&h=0000000000 - no "iss" → "i"

**Comment by ekaf** [comment]
Yes @novalis, a lexicon of inflected forms would be nice to have. There exists a good one for French, [The Lefff, a Freely Available and Large-coverage Morphological and Syntactic Lexicon for French](https://aclanthology.org/L10-1487/). However, such resources can only handle words within the limits of the vocabulary that they know.

**Comment by ekaf** [comment]
Thanks @caiw and @nezda: indeed, the original Morphy from Princeton does not have these bugs, and neither does my Prolog implementation, [wn_morphy.pl](https://github.com/ekaf/wordnet-prolog/blob/master/wn_morphy.pl).
So there must be something wrong with the implementation of morphy in NLTK's wordnet.py module, and we need to worry about this issue.

**Comment by ekaf** [comment]
_WordNetLemmatizer_ and _morphy_ are only simple wrappers, that pick just one single lemma from the list of lemmas found by _\_morphy_.  The strategies chosen (shortest  vs. first found lemma) may not always be adequate, and it is especially problematic that _WordNetLemmatizer_ just returns any garbage input as its own lemma. However, I don't think these wrappers need fixing. Instead,  alternative stemmers are available, and more can be developed.

But we can, and should, fix the main _\_morphy_ function in _wordnet.py_, which returns a list of lemma candidates. It is mostly a faithful implementation of Princeton Morphy, except from one huge difference: whereas the original only does one pass over the possible morphological substitutions, the NLTK implementation can recursively keep removing suffixes. This practice dates back to the original _pywordnet_ by @osteele, but it is not a part of the original algorithm from Princeton Morphy. 

For example, without this spurious recursion step, no lemma would be found for "iss" using _morphy_, because the "s" ending would only be stripped once.

**Comment by nezda** [comment]
now that's what i call a deep dive - nice one @ekaf
